test/tests/functional/pbs_qrun.py

```diff
@@ -92,3 +93,39 @@ def test_invalid_host_val(self):
         msg = "Server is not up"
         self.assertTrue(self.server.isUp(), msg)
         self.logger.info("As expected server is up and running")
+
+    def test_qrun_hangs(self):
+        """
+        This test verifies whether qrun hangs, if scheduling is
+        disabled
+        """
+        node = self.mom.shortname
+        time1 = int(time.time())
+        while True:
+            self.server.manager(MGR_CMD_SET, NODE,
+                                {'resources_available.ncpus': 1}, id=node)
+            j1 = Job(TEST_USER)
+            j1.create_script('/bin/date')
+            self.server.submit(j1)
+            jid2 = self.server.submit(j1)
+
+            self.server.manager(MGR_CMD_SET, SCHED,
+                                {'scheduling': 'True'})
+            self.server.manager(MGR_CMD_SET, SCHED,
+                                {'scheduling': 'False'})
```

Comment:
Add expect=True.  It will wait for the current cycle to end before returning.

Reply:
Actually as per the shell script, the bug is reproducible only when that qrun job is processed in existing cycle and there are no free nodes available to run the job.
So, i guess we need the current cycle on-going, when qrun process the job.

Reply:
That gets tricky.  Since there is no real way to force the scheduler to stay in cycle, you'll need to submit a ton of jobs to cause the scheduler to do a lot of work.  They need to be different jobs or the scheduler will just ignore them after it realized it can't run one in the job equivalence class.

Reply:
Yes @bhroam , if we look at the shell script, where the bug was reported and also able to re-produce, there we need to submit different jobs infinitely in a while loop, to hit that particular scenario.
While converting the shell script to PTL test case, we need to submit jobs infinitely (but in PTL, I submitted job till 1 min) and the issue was reproducable (as the qrun hangs), in some cycle. Attached the log above.

Reply:
race conditions are tricky to reproduce in automated tests.  I may have been amiss asking you to write one.  I didn't understand fully what you had to do (sorry).  That said, I think we can craft a test that does what you need.  Turn scheduling off.  Submit 500-1000 different jobs.  The easiest way to do this is to increment the walltime by 1s for each job.  This will force the scheduler to put each job into its own equivalence class.  Turn scheduling on, and the cycle will take several seconds.  During this time you can do your qrun.  I worry doing what you are doing will fail on different speed machines.